src/c_api/adapter.h

```diff
@@ -0,0 +1,478 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file adapter.h
+ */
+#ifndef XGBOOST_C_API_ADAPTER_H_
+#define XGBOOST_C_API_ADAPTER_H_
+#include <limits>
+#include <memory>
+#include <string>
+namespace xgboost {
+
+/**  External data formats should implement an adapter as below. The
+ * adapter provides a uniform access to data outside xgboost, allowing
+ * construction of DMatrix objects from a range of sources without duplicating
+ * code.
+ *
+ * The adapter object is an iterator that returns batches of data. Each batch
+ * contains a number of "lines". A line represents a set of elements from a
+ * sparse input matrix, normally a row in the case of a CSR matrix or a column
+ * for a CSC matrix. Typically in sparse matrix formats we can efficiently
+ * access subsets of elements at a time, but cannot efficiently lookups elements
+ * by random access, hence the "line" abstraction, allowing the sparse matrix to
+ * return subsets of elements efficiently. Individual elements are described by
+ * a COO tuple (row index, column index, value).
+ *
+ * This abstraction allows us to read through different sparse matrix formats
+ * using the same interface. In particular we can write a DMatrix constructor
+ * that uses the same code to construct itself from a CSR matrix, CSC matrix,
+ * dense matrix or libsvm file. To see why this is necessary, imagine we have 5
+ * external matrix formats and 5 internal DMatrix types where each DMatrix needs
+ * a custom constructor for each possible input. The number of constructors is
+ * 5*5=25. Using an abstraction over the input data types the number of
+ * constructors is reduced to 5, as each DMatrix is oblivious to the external
+ * data format. Adding a new input source is simply a case of implementing an
+ * adapter.
+ *
+ * Most of the below adapters do not need more than one batch as the data
+ * originates from an in memory source. The file adapter does require batches to
+ * avoid loading the entire file in memory.
+ *
+ * An important detail is empty row/column handling. Files loaded from disk do
+ * not provide meta information about the number of rows/columns to expect, this
+ * needs to be inferred during construction. Other sparse formats may specify a
+ * number of rows/columns, but we can encounter entirely sparse rows or columns,
+ * leading to disagreement between the inferred number and the meta-info
+ * provided. To resolve this, adapters have methods specifying the number of
+ * rows/columns expected, these methods may return zero where these values must
+ * be inferred from data. A constructed DMatrix should agree with the input
+ * source on numbers of rows/columns, appending empty rows if necessary.
+ *  */
+
+struct COOTuple {
```

Comment:
Do you have an expected effect on memory usage now that we have an intermediate type?

Reply:
This type is only created when we access elements one at a time, the data is never stored using this type.